grammars/silver/compiler/extension/records/Records.sv.md

```diff
@@ -0,0 +1,370 @@
+
+```silver example
+nonterminal Foo;
+abstract production foo
+Foo ::=
+{}
+
+annotation field2<a>::a;
+annotation field1<Integer> occurs on Foo;
+annotation field2<String> occurs on Foo;
+
+nonterminal Bar;
+abstract production bar
+Bar ::=
+{}
+
+annotation field1<a>::a;
+annotation field1<Boolean> occurs on Bar;
+annotation field2<Float> occurs on Bar;
+```
+
+Type inference should always be able to sort out what the type of `thing.field1` is correctly, although we don't currently have a way to talk about "types with a `field1` with type `String`."
+
+```silver
+synthesized attribute recordFieldAGDcl::AGDcl;
+
+nonterminal RecordField with location, unparse, recordFieldAGDcl,
+  recordFieldAGDclQName, withoutDeclaringInh;
+
+concrete production recordField
+this::RecordField ::= name::IdLower_t  '::'  typeExpr::TypeExpr
+{
+  this.unparse = s"${name.lexeme}::${typeExpr.unparse}";
+
+  local qname::QName =
+    qNameId(nameIdLower(name, location=this.location), location=this.location);
+
+  local typeVarExpr::TypeExpr =
+    typeVariableTypeExpr(terminal(IdLower_t, "a", this.location),
+      location=this.location);
+  local typeVarArgs::BracketedOptTypeExprs =
+    botlSome(
+      bTypeList('<', typeListSingle(typeVarExpr, location=this.location), '>',
+        location=this.location),
+      location=this.location);
+
+  local typeArgs::BracketedOptTypeExprs =
+    botlSome(
+      bTypeList('<', typeListSingle(typeExpr, location=this.location), '>',
+        location=this.location),
+      location=this.location);
+
+  local declare::AGDcl =
+    annotationDcl('annotation', qname, typeVarArgs, '::', typeVarExpr, ';',
+      location=this.location);
+  local occur::AGDcl =
+    annotateDcl('annotation', qname, typeArgs, 'occurs', 'on',
+      this.recordFieldAGDclQName, botlNone(location=this.location), ';',
+      location=this.location);
+
+  this.recordFieldAGDcl =
+    if contains(name.lexeme, this.withoutDeclaringInh)
+    then occur
+    else appendAGDcl(declare, occur, location=this.location);
+}
+```
+
+Note that in order to know the name of the nonterminal the attributes should occur on, we pass it down as an inherited attribute.
+
+```silver
+inherited attribute recordFieldAGDclQName::QName;
+
+aspect production recordDecl
+this::AGDcl ::= quals::NTDeclQualifiers  'record'  name::IdUpper_t
+  modifiers::NonterminalModifiers  '{'  fields::RecordFields0  '}'  ';'
+{
+  fields.recordFieldAGDclQName =
+    qNameId(nameIdUpper(name, location=this.location), location=this.location);
+}
+```
+
+### Field Lists
+
+Sequences of fields exist in a pretty typical way.
+There can be zero fields, a single field, or a series of fields; fields are comma-separated, and a trailing comma is allowed.
+The grammar is:
+
+```
+RecordFields ::=
+              |  RecordField
+              |  RecordField ',' RecordFields
+```
+
+However, for parsing efficiency, we want to make this grammar left-recursive instead of right-recursive.
```

Comment:
Is this actually a measurable performance impact?  Also I think this grammar allows a trailing comma, was that intended?

Reply:
> performance impact

From this extension alone, no, but my understanding was that in general, you want to make as many things as possible left-recursive because it matters for parsing perf.

> trailing comma

Yeah, that was intentional; I typically write records with trailing commas in other languages because then I don't need to worry about it when changing or reordering fields, and diffs don't have the extra "noise" of:

```diff
 record Foo {
     bar: String,
-    baz: String
+    baz: String,
+    quux: String
 }
```

Reply:
Ah yes, makes sense.